spec/src/main/resources/META-INF/smithy/bsp/bsp.smithy

```diff
@@ -0,0 +1,1380 @@
+$version: "2"
+
+namespace bsp
+
+use jsonrpc#enumKind
+use jsonrpc#jsonNotification
+use jsonrpc#jsonRPC
+use jsonrpc#jsonRequest
+
+@jsonRPC
+service BuildServer {
+    operations: [
+        InitializeBuild
+        OnBuildInitialized
+        ShutdownBuild
+        OnBuildShutdown
+        WorkspaceBuildTargets
+        WorkspaceReload
+        BuildTargetSources
+        BuildTargetInverseSources
+        BuildTargetDependencySources
+        BuildTargetDependencyModules
+        BuildTargetResources
+        BuildTargetOutputPaths
+        BuildTargetCompile
+        BuildTargetRun
+        DebugSessionStart
+        BuildTargetCleanCache
+    ]
+}
+
+@jsonRPC
+service BuildClient {
+    operations: [
+        OnBuildShowMessage
+        OnBuildLogMessage
+        OnBuildPublishDiagnostics
+        OnBuildTargetDidChange
+        OnBuildTaskStart
+        OnBuildTaskProgress
+        OnBuildTaskFinish
+    ]
+}
+
+/// Like the language server protocol, the initialize request is sent as the first request from the client to the server.
+/// If the server receives a request or notification before the initialize request it should act as follows:
+///
+/// * For a request the response should be an error with code: -32002. The message can be picked by the server.
+/// * Notifications should be dropped, except for the exit notification. This will allow the exit of a server without an initialize request.
+///
+/// Until the server has responded to the initialize request with an InitializeBuildResult, the client must not send any additional
+/// requests or notifications to the server.
+@jsonRequest("build/initialize")
+operation InitializeBuild {
```

Comment:
I'm not really sure how to phrase the question, so bear with me. Looking at the docs and such I see the `operation`s can also have `errors`. In this scenario where the server receives a request or notification before the initialize and the response is supposed to be a `-32002`, is that able to be captured at all in the spec here? I know it doesn't belong in `InitializeBuild.errors`, but is there anything that can capture that?

Reply:
It'd certainly be possible to define a `jsonRPCError` trait, use it to annotate structures and list those structures in operations. Then, code generators could potentially generate those structures as throwable an apply some special handling in the routing logic to apply the correct error code when such throwable are caught. 

However, this is way beyond the scope of this PR, the goal of which is to try and capture the information present in the existing specification as opposed to improving / filling the gaps of the existing specification.

Reply:
Gotcha, makes sense. Thanks!